vignettes/in-packages.Rmd

```diff
@@ -0,0 +1,399 @@
+
+If you use Travis-CI already, we **strongly** recommend adding a build that tests with the development version of tidyr; see above for details.
+
+This section briefly describes how to run different code for different versions of tidyr, then goes through the major changes that might require workarounds: 
+
+* `nest()` and `unnest()` get new interfaces.
+* `nest()` preserves groups.
+* `nest_()` and `unnest_()` are defunct.
+
+If you're struggling with a problem that's not described here, please reach out via [github](https://github.com/tidyverse/tidyr/issues/new) or [email](hadley@rstudio.com) so we can help out.
+
+### Conditional code
+
+Sometimes you'll be able to write code that works with v0.8.3 _and_ v1.0.0. But this often requires code that's not particularly natural for either version and you'd be better off to (temporarily) have separate code paths, each containing non-contrived code. You get to re-use your existing code in the "old" branch, which will eventually be phased out, and write clean, forward-looking code in the "new" branch.
+
+The basic approach looks like this. First you define a function that returns `TRUE` for new versions of tidyr:
+
+```{r}
+tidyr_new_interface <- function() {
+  packageVersion("tidyr") > "0.8.99"
+}
+```
+
+We highly recommend keeping this as a function because it provides an obvious place to jot any transition notes for your package, and it makes it easier to remove transitional code later on. Another benefit is that the tidyr version is determined at *run time*, not at *build time*, and will therefore detect your user's current tidyr version.
+
+Then in your functions, you use an `if` statement to call different code for different versions:
+
+```{r, eval = FALSE}
+my_function_inside_a_package <- function(...)
+  # my code here
+
+  if (tidyr_new_interface()) {
+    # Freshly written code for v1.0.0
+    out <- tidyr::nest(df, data = one_of("x", "y", "z"))
+  } else {
+    # Existing code for v0.8.3
+    out <- tidyr::nest(df, x, y, z)
+  }
+
+  # more code here
+}
+```
+
+If your new code uses a function that only exists in tidyr 1.0.0, you will get a `NOTE` from `R CMD check`: this is one of the few notes that you can explain in your CRAN submission comments. Just mention that it's for forward compatibility with tidyr 1.0.0, and CRAN will let your package through.
+
+### New syntax for `nest()`
+
+What changed:
+
+*   The to-be-nested columns are no longer accepted as "loose parts".
+*   The new list-column's name is no longer provided via the `.key` argument.
+*   Now we use a construct like this: `new_col = <something about existing cols>`.
+  
+Why it changed:
+
+*   The use of `...` to make data is a problematic pattern we're moving away from.
```

Comment:
I think `...` to make data is exactly when it's not problematic. It's the use of `...` for metadata (e.g. a selection) that's problematic, IIUC.